File: dynet-py/rnnlm.py
from collections import Counter, defaultdict
from itertools import count
import random
import time
import math
import sys

import dynet as dy
import numpy as np

# format of files: each line is "word1/tag2 word2/tag2 ..."
train_file="data/text/train.txt"
test_file="data/text/dev.txt"

w2i = defaultdict(count(0).next)

def read(fname):
    """
    Read a file where each line is of the form "word1 word2 ..."
    Yields lists of the form [word1, word2, ...]
    """
    with file(fname) as fh:
        for line in fh:
            sent = [w2i[x] for x in line.strip().split()]
            sent.append(w2i["<s>"])
            yield sent

train=list(read(train_file))
nwords = len(w2i)
test=list(read(test_file))
S = w2i["<s>"]
assert(nwords == len(w2i))

# DyNet Starts

model = dy.Model()
trainer = dy.AdamTrainer(model)

# Lookup parameters for word embeddings
WORDS_LOOKUP = model.add_lookup_parameters((nwords, 64))

# Word-level LSTM (layers=1, input=64, output=128, model)
RNN = dy.LSTMBuilder(1, 64, 128, model)

# Softmax weights/biases on top of LSTM outputs
W_sm = model.add_parameters((nwords, 128))
b_sm = model.add_parameters(nwords)

# Build the language model graph
def calc_lm_loss(sent):

    dy.renew_cg()
    # parameters -> expressions
    W_exp = dy.parameter(W_sm)
    b_exp = dy.parameter(b_sm)

    # initialize the RNN
    f_init = RNN.initial_state()

    # start the rnn by inputting "<s>"
    s = f_init.add_input(WORDS_LOOKUP[sent[-1]]) 

    # feed word vectors into the RNN and predict the next word
    losses = []
    for wid in sent:
        # calculate the softmax and loss
        score = W_exp * s.output() + b_exp
        loss = dy.pickneglogsoftmax(score, wid)
        losses.append(loss)
        # update the state of the RNN
        s = s.add_input(WORDS_LOOKUP[wid]) 
    
    return dy.esum(losses)

start = time.time()
i = all_time = all_tagged = this_words = this_loss = 0
for ITER in xrange(10):
    random.shuffle(train)
    for s in train:
        i += 1
        if i % 500 == 0:
            trainer.status()
            print this_loss / this_words
            all_tagged += this_words
            this_loss = this_words = 0
        if i % 5000 == 0:
            all_time += time.time() - start
            dev_loss = dev_words = 0
            for sent in test:
                loss_exp = calc_lm_loss(sent)
                dev_loss += loss_exp.scalar_value()
                dev_words += len(sent)
            print ("nll=%.4f, ppl=%.4f, time=%.4f, word_per_sec=%.4f" % (dev_loss/dev_words, math.exp(dev_loss/dev_words), all_time, all_tagged/all_time))
            if all_time > 3600:
                sys.exit(0)
            start = time.time()
        # train on sent
        loss_exp = calc_lm_loss(s)
        this_loss += loss_exp.scalar_value()
        this_words += len(s)
        loss_exp.backward()
        trainer.update()
    trainer.update_epoch(1.0)

File: dynet-cpp/rnnlm.cc
#include <string>
#include <vector>
#include <sstream>
#include <fstream>
#include <stdexcept>
#include <algorithm>
#include <unordered_map>
#include <chrono>

#include <dynet/dict.h>
#include <dynet/expr.h>
#include <dynet/lstm.h>
#include <dynet/training.h>

using namespace std;
using namespace std::chrono;
using namespace dynet;
using namespace dynet::expr;

// Read a file where each line is of the form "word1 word2 ..."
// Yields lists of the form [word1, word2, ...]
vector<vector<int> > read(const string & fname, Dict & vw) {
  ifstream fh(fname);
  if(!fh) throw std::runtime_error("Could not open file");
  string str; 
  vector<vector<int> > sents;
  while(getline(fh, str)) {
    istringstream iss(str);
    vector<int> tokens;
    while(iss >> str)
      tokens.push_back(vw.convert(str));
    tokens.push_back(vw.convert("<s>"));
    sents.push_back(tokens);
  }
  return sents;
}

struct RNNLanguageModel {
  LookupParameter p_c;
  Parameter W_sm;
  Parameter b_sm;
  VanillaLSTMBuilder builder;
  explicit RNNLanguageModel(unsigned layers, unsigned input_dim, unsigned hidden_dim, unsigned vocab_size, Model& model) : builder(layers, input_dim, hidden_dim, model) {
    p_c = model.add_lookup_parameters(vocab_size, {input_dim}); 
    W_sm = model.add_parameters({vocab_size, hidden_dim});
    b_sm = model.add_parameters({vocab_size});
  }

  Expression calc_lm_loss(const vector<int> & sent, ComputationGraph & cg) {
  
    // parameters -> expressions
    Expression W_exp = parameter(cg, W_sm);
    Expression b_exp = parameter(cg, b_sm);
  
    // initialize the RNN
    builder.new_graph(cg);  // reset RNN builder for new graph
    builder.start_new_sequence();
  
    // start the rnn by inputting "<s>"
    Expression s = builder.add_input(lookup(cg, p_c, *sent.rbegin())); 
  
    // feed word vectors into the RNN and predict the next word
    vector<Expression> losses;
    for(auto wid : sent) {
      // calculate the softmax and loss
      Expression score = affine_transform({b_exp, W_exp, s});
      Expression loss = pickneglogsoftmax(score, wid);
      losses.push_back(loss);
      // update the state of the RNN
      s = builder.add_input(lookup(cg, p_c, wid));
    }
    
    return sum(losses);
  }

};

int main(int argc, char** argv) {

  // format of files: each line is "word1 word2 ..."
  string train_file = "data/text/train.txt";
  string test_file = "data/text/dev.txt";

  Dict vw;
  vector<vector<int> > train = read(train_file, vw);
  vw.freeze();
  vector<vector<int> > test = read(test_file, vw);

  int nwords = vw.size();

  // DyNet Starts
  dynet::initialize(argc, argv);
  Model model;
  AdamTrainer trainer(model, 0.001);
  trainer.sparse_updates_enabled = false;

  RNNLanguageModel rnnlm(1, 64, 128, nwords, model);

  time_point<system_clock> start = system_clock::now();
  int i = 0, all_tagged = 0, this_words = 0;
  float this_loss = 0.f, all_time = 0.f;
  for(int iter = 0; iter < 10; iter++) {
    shuffle(train.begin(), train.end(), *dynet::rndeng);
    for(auto & s : train) {
      i++;
      if(i % 500 == 0) {
        trainer.status();
        cout << this_loss/this_words << endl;
        all_tagged += this_words;
        this_loss = 0.f;
        this_words = 0;
      }
      if(i % 5000 == 0) {
        duration<float> fs = (system_clock::now() - start);
        all_time += duration_cast<milliseconds>(fs).count() / float(1000);
        int test_words = 0;
        float test_loss = 0;
        for(auto & sent : test) {
          ComputationGraph cg;
          Expression loss_exp = rnnlm.calc_lm_loss(sent, cg);
          test_loss += as_scalar(cg.forward(loss_exp));
          test_words += sent.size();
        }
        cout << "nll=" << test_loss/test_words << ", ppl=" << exp(test_loss/test_words) << ", words=" << test_words << ", time=" << all_time << ", word_per_sec=" << all_tagged/all_time << endl;
        if(all_time > 3600)
          exit(0);
        start = system_clock::now();
      }

      ComputationGraph cg;
      Expression loss_exp = rnnlm.calc_lm_loss(s, cg);
      this_loss += as_scalar(cg.forward(loss_exp));
      this_words += s.size();
      cg.backward(loss_exp);
      trainer.update();
    }
    trainer.update_epoch(1.0);
  }
}
